import config

if config.ref == "mm10":
    bwa_ref = "/installed_tools/RefGenomes/mm10_all/BWA/mm10_ordered_all.fa"
    library = "SureSelect_XT_Mouse_All_Exome"
    refgen = "/installed_tools/RefGenomes/mm10_all/BOWTIE2/mm10_ordered_all"
    bwa_ref = "/installed_tools/RefGenomes/mm10_all/BWA/mm10_ordered_all.fa"
    org = "MOUSE"
    target_bed = "/installed_tools/RefGenomes/exome/mouse/S0276129_Regions_mm10.bed6"	
    refflat_file = "/installed_tools/RefGenomes/mm10_GTF/gencode.vM9.refFlat.txt"
    rRNA_interval = "/installed_tools/RefGenomes/mm10_GTF/gencode.vM9.rRNA.interval"
    rseqc_file = "/installed_tools/RefGenomes/mm10_GTF/RSeQC_GeneModel/GRCm38_mm10_Ensembl.bed"
    star_genome = "/installed_tools/RefGenomes/mm10_all/RSEM/mm10_genM9_star_rsem_overhang125"
    gtf_file = "/installed_tools/RefGenomes/mm10_GTF/gencode.vM9.annotation.gtf"
    rsem_transcrtiptome = "/installed_tools/RefGenomes/mm10_all/RSEM/mm10_genM9_star_rsem_overhang125/mm10_genM9_star_rsem"
    salmon_transcrtiptome = "/installed_tools/RefGenomes/mm10_all/kallisto/GRCm38_gencodeM9/GRCm38_gencodeM9"	
	
if config.ref == "mm9":
    refgen = "/installed_tools/RefGenomes/mm9_bowtie2/mm9"
    star_genome = "/installed_tools/RefGenomes/STAR/mm9_STAR_with_ensembl_NCBIM37_annotaion_overhang125"
    refflat_file = "/installed_tools/RefGenomes/mm9_mRNASplice/RefFlat/refFlat.txt"
    rRNA_interval = "/installed_tools/RefGenomes/mm9_GTF/RSeQC_GeneModel/mm9_NCBI37_Ensembl.bed"
    rseqc_file = "/installed_tools/RefGenomes/mm9_GTF/RSeQC_GeneModel/mm9_NCBI37_Ensembl.bed"
    gtf_file = "/installed_tools/RefGenomes/mm9_GTF/mm9_ensembl_ncbim37_igenome_chr.gtf"
    rsem_transcrtiptome = ""
    salmon_transcrtiptome = ""	
	
if config.ref == "hg19":
    bwa_ref = "/installed_tools/RefGenomes/hg19_BWA/hg19_chrM_1st.fa"
    org = "HUMAN"
    target_bed = "/installed_tools/RefGenomes/exome/S04380219_V5-UTR_covered.bed6"
    library = "SureSelect_hg_all_exon_v5_utr_exome"
    refgen = "/installed_tools/RefGenomes/hg19_bowtie2/hg19"
    refflat_file = "/installed_tools/RefGenomes/hg19_GTF/hg19_Ensembl_v70_refFlat.txt"
    rRNA_interval = "/installed_tools/RefGenomes/hg19_GTF/star_GRCh37_ribosome_interval_list.txt"
    rseqc_file = "/installed_tools/RefGenomes/hg19_GTF/RSeQC_GeneModel/hg19_Ensembl.bed"
    gtf_file = "/installed_tools/RefGenomes/hg19_GTF/hg19_Ensembl_Gene_v70.GTF"
    star_genome = "/installed_tools/RefGenomes/RSEM/RSEM_star_hg19_ens70_overhang125/"
    rsem_transcrtiptome = "/installed_tools/RefGenomes/RSEM/RSEM_star_hg19_ens70_overhang125/hg19_ens70_star_rsem"
    salmon_transcrtiptome = "/installed_tools/RefGenomes/hg19_cDNA/Homo_sapiens.GRCh37.70.cdna.all.ed.name.fa"
    fregs="/installed_tools/ancestry/hapmap3.8populations.hg19.freqs"
	fai = "/installed_tools/RefGenomes/hg19/fasta/hg19_genome_with_contigs.fa.fai"
	svviz_gtf="/installed_tools/RefGenomes/hg19/GTF/gencode.v24.annotation.sort.gtf.gz"
		
if config.ref == "hg38":
    bwa_ref = "/installed_tools/RefGenomes/hg38/BWA/hg38_genome_with_contigs.fa"
    bwa_seqc = "/installed_tools/RefGenomes/hg38_SEQC/BWA/hg38.fasta"
	org = "HUMAN"
    target_bed = "/installed_tools/RefGenomes/exome/S04380219_Covered_liftover_to_hg38.bed6"
    library = "SureSelect_hg_all_exon_v5_utr_exome"
    refgen = "/installed_tools/RefGenomes/hg38/BOWTIE2/hg38"
    bwa_ref = "/installed_tools/RefGenomes/hg38/BWA/hg38_genome_with_contigs.fa"
    refflat_file = "/installed_tools/RefGenomes/hg38/Picard_and_RSeqc/gencode.v24.rRNA.refFlat.txt"
    rRNA_interval = "/installed_tools/RefGenomes/hg38/Picard_and_RSeqc/gencode.v24.rRNA.interval_list"
    rseqc_file = "/installed_tools/RefGenomes/hg38/Picard_and_RSeqc/RSeqc.GRCh38.79.bed"
    gtf_file = "/installed_tools/RefGenomes/hg38/GTF/gencode.v24.annotation.gtf"
    star_genome = "/installed_tools/RefGenomes/hg38/RSEM/RSEM_star_grch38_gencode24_overhang125"
    rsem_transcrtiptome = "/installed_tools/RefGenomes/hg38/RSEM/RSEM_star_grch38_gencode24_overhang125/hg38_gen24_star_rsem"
    salmon_transcrtiptome = "/installed_tools/RefGenomes/hg38/cDNA_transcriptome/gencode.v24.transcripts.single.name.fa"
	melt1 = "/installed_tools/Software/tools/MELT/MELTv2.1.5/me_refs/Hg38/list.txt"
	melt2 = "/installed_tools/Software/tools/MELT/MELTv2.1.5/add_bed_files/Hg38/Hg38.genes.bed"
	msisensor_filter = "/installed_tools/RefGenomes/hg38_SEQC/hg38.filter.msisensor.txt"
	fregs="/installed_tools/ancestry/hapmap3.8populations.hg38.freqs"
	fai = "/installed_tools/RefGenomes/hg38/fasta/hg38_genome_with_contigs.fa.fai"
	svviz_gtf="/installed_tools/RefGenomes/hg38/GTF/gencode.v24.annotation.sort.gtf.gz"